File: KeyChain/KeyEntity.m
    //
//  KeyEntity.m
//  KeyChain
//
//  Created by softphone on 10/01/11.
//  Copyright 2011 SOUL. All rights reserved.
//

#import "KeyEntity.h"


static NSString * _MNEMONIC = @"mnemonic";
static NSString * _IS_NEW = @"isNew";

@implementation KeyEntity

@synthesize sectionId;
@dynamic mnemonic;
@dynamic group, groupPrefix;


#pragma mark -
#pragma mark Grouping section
#pragma mark -

//
- (void)groupByPrefix:(NSString*)prefix {
    
    if( prefix == nil ) return;

    self.groupPrefix = prefix;
    self.group = [NSNumber numberWithBool:YES];
    
}

//
- (void)detachFromGroup {

    self.groupPrefix = nil;
    self.group = [NSNumber numberWithBool:NO];
    
}


/*
//
+ (void)groupByAppendingPrefix:(KeyEntity *)key prefix:(NSString*)prefix;
{
    
    if (prefix==nil || key==nil) {
        return;
    }
    
    key.mnemonic = [prefix stringByAppendingString:key.mnemonic];
    key.group = [NSNumber numberWithBool:YES];
}

//
+ (void)groupByReplacingName:(KeyEntity *)key mnemonic:(NSString*)name {
    if (name==nil || key==nil) {
        return;
    }
    
    key.mnemonic = name;
    key.group = [NSNumber numberWithBool:YES];
    
}

//
+ (void)groupByReplacingPrefix:(KeyEntity *)key groupKey:(NSString*)groupKey prefix:(NSString*)prefix {
    
    NSString *newMnemonic = [key.mnemonic 
                             stringByReplacingOccurrencesOfString:groupKey 
                             withString:prefix 
                             options:NSCaseInsensitiveSearch 
                             range:NSMakeRange(0, [groupKey length])];
    key.mnemonic = newMnemonic;
    key.group = [NSNumber numberWithBool:YES];
    
}
*/

#pragma mark static implementation

static  NSString * _REGEXP = @"(\\w+[-@/])(\\w+)";


+ (BOOL)isSectionAware:(KeyEntity *)key {
    
    NSPredicate *predicate = 
        [NSPredicate predicateWithFormat:@"SELF.mnemonic MATCHES %@",  _REGEXP];
    
    BOOL result = [predicate evaluateWithObject:key];
    
    return result;
}

+ (NSRange)getSectionPrefix:(KeyEntity*)key checkIfIsSectionAware:(BOOL)check {
    
    //@autoreleasepool {
        
        NSError *error = nil;
    
        NSRange result = NSMakeRange(NSNotFound, 0);

        if( !check || [KeyEntity isSectionAware:key] ) {
            
            NSRegularExpression *pattern = [[NSRegularExpression alloc] 
                                             initWithPattern:_REGEXP 
                                             options:NSRegularExpressionCaseInsensitive 
                                             error:&error ] ;
            if (pattern!=nil) {
                NSTextCheckingResult *match = 
                    [pattern firstMatchInString:key.mnemonic 
                                    options:0 
                                      range:NSMakeRange(0, [key.mnemonic length])];
                
                
                
                result = [match rangeAtIndex:1];
                
            }
            
        }

        return result;
    //}
    
}

+ (NSString *)sectionNameFromPrefix:(NSString *)prefix trim:(BOOL)trim {

    if( prefix == nil ) return nil;
    
    if( trim ) {
        prefix = [prefix stringByTrimmingCharactersInSet:[NSCharacterSet whitespaceAndNewlineCharacterSet]];
    }
    
    if( [prefix length] == 0 ) {
        return nil;
    }

    NSString *result = [prefix substringToIndex:[prefix length] -1];
    
    return result;
}


+ (KeyEntity *)createSection:  (NSString *)groupKey 
                  groupPrefix:(NSString *)groupPrefix
                    inContext:(NSManagedObjectContext *)context {
    NSEntityDescription *entity = [NSEntityDescription entityForName:@"KeyInfo" inManagedObjectContext:context];

   // NSString *entityName = [[source entity] name];
    NSString *entityName = [entity name];
    
    //create new object in data store
    KeyEntity *cloned = [NSEntityDescription
                         insertNewObjectForEntityForName:entityName
                         inManagedObjectContext:context];

/*    
    //loop through all attributes and assign then to the clone
    NSDictionary *attributes = [[NSEntityDescription
                                 entityForName:entityName
                                 inManagedObjectContext:context] attributesByName];
    
    for (NSString *attr in attributes) {
        if ([attr compare:_MNEMONIC]==0 ) continue;
        if ([attr compare:_IS_NEW]==0 ) continue;
        
        [cloned setValue:[source valueForKey:attr] forKey:attr];
        
    }
*/   
    [cloned setValue:@"nil" forKey:@"password"];
    [cloned setValue:@"nil" forKey:@"username"];
    
    cloned.group = NO;
    cloned.mnemonic = groupKey;
    cloned.groupPrefix = groupPrefix;
    
    return cloned;     
}


#pragma mark -
#pragma mark implementation
#pragma mark -

- (BOOL)isSection {
    return self.groupPrefix != nil && (self.group == nil || [self.group boolValue] == NO);
}

- (BOOL)isGrouped {
    return (self.groupPrefix != nil && self.group != nil && [self.group boolValue] == YES);
}

- (BOOL)isEqualForImport:(id)object {

    if( object!=nil ) {
        
        NSString *k1 = nil;
    
        if ( [object isKindOfClass:[KeyEntity class]] ) {
        
            k1 = ((KeyEntity *)object).mnemonic;
        
        }
        else if( [object isKindOfClass:[NSDictionary class]] ) {
            
            k1 = [object valueForKey:_MNEMONIC];
            
        }
        else if( [object isKindOfClass:[NSString class]] ) {
            
            k1 = object ;
            
        }
    
        NSString *k = self.mnemonic;
        
        return (k!=nil) ? [k isEqualToString:k1] : NO;

    }
    
    return NO;
    
}

- (void)awakeFromFetch {

    [self setPrimitiveValue:[NSNumber numberWithBool:NO] forKey:_IS_NEW];
	
}

- (void)didSave {
    
    
    [self setPrimitiveValue:[NSNumber numberWithBool:NO] forKey:_IS_NEW];
}

//- (NSString *)mnemonic { return [self valueForKey:@"mnemonic"]; }
//- (void)setMnemonic:(NSString*)value { [self setValue:value forKey:@"mnemonic"]; }


-(BOOL)isNew {
	
	BOOL value = [[self primitiveValueForKey:_IS_NEW] boolValue];
	NSLog( @"isNew [%d]", value );
	
	return value;
	

}

- (NSString *)sectionId {

	NSString *k = [self valueForKey:_MNEMONIC];
	
	return [k substringWithRange:NSMakeRange( 0, 1)];
}


#pragma mark - Serialization

- (NSDictionary *)toDictionary:(NSMutableDictionary*)target {
    
    if (target == nil ) {
        return target;
    }
    
    NSArray* attributes = [[[self entity] attributesByName] allKeys];
    
    for (NSString *key in attributes) {
        
        if ([key compare:_IS_NEW]==0 ) continue;
        
        NSObject* value = [self valueForKey:key];
        
        if (value != nil) {
            [target setObject:value forKey:key];
        }
        
    }
    return target;
}

- (void)fromDictionary:(NSDictionary *)source {

    if (source == nil ) {
        return;
    }
    
    NSArray* attributes = [[[self entity] attributesByName] allKeys];
    
    for (NSString *key in attributes) {
        
        if ([key compare:_IS_NEW]==0 ) continue;
        
        NSObject* value = [source valueForKey:key];
        
        if (value != nil) {
            [self setValue:value forKey:key];
        }
        
    }
}

@end

File: KeyChain/KeyEntity.swift
//
//  KeyEntity.swift
//  KeyChain
//
//  Created by softphone on 28/12/15.
//  Copyright © 2015 SOFTPHONE. All rights reserved.
//

import Foundation
import CoreData
import KeychainAccess

@objc class StringEncryptionTransformer : ValueTransformer {

    override class func transformedValueClass() -> AnyClass {
        return Data.self as! AnyClass
    }
    override func transformedValue(_ value: Any?) -> Any? // by default returns value
    {
        return value
    }
    
    override func reverseTransformedValue(_ value: Any?) -> Any? // by default raises an exception if +allowsReverseTransformation
    {
        return value
    }
}

@objc class KeyEntity : NSManagedObject {
    
    @NSManaged var mnemonic:String
    @NSManaged var groupPrefix:String?
    @NSManaged var group:NSNumber?
    @NSManaged var note:String
    
    
    
    static let _MNEMONIC    = "mnemonic"
    static let _PASSWORD    = "password"
    static let _IS_NEW      = "isNew"

    //@TRANSIENT
    var isNew:Bool {
    
        get {
            let value = (self.primitiveValue(forKey: KeyEntity._IS_NEW)! as AnyObject).boolValue
            
            return value!;
            
        }
    
    }
    
    //@TRANSIENT
    var password:String? {
        
        get {
            let keychain = Keychain(service: AccountCredential.sharedCredential.bundleId)
            
            let key = self.mnemonic
            
            if let data =  try! keychain.getString(key) {
                return data
            }
            
            return nil
        }
        set(value) {
            
            if let v = value {
                let keychain = Keychain(service: AccountCredential.sharedCredential.bundleId)
 
                let key = self.mnemonic
                
                try! keychain
                    .accessibility(.whenUnlocked)
                    .set(v, key: key)
            }
        }
    }
    
    var sectionId:String? {
    
        get {
            if let k = self.value(forKey: KeyEntity._MNEMONIC) {
            
                return (k as AnyObject).substring(with: NSMakeRange(0,1))
            }
            
            return nil
        }
    }
    

    //MARK: Grouping section
    
    func groupByPrefix( _ prefix:String? ) -> Void {
    
        if let p = prefix  {
            self.groupPrefix = p;
            self.group = NSNumber(value: true as Bool)
        }
    }
    
    func detachFromGroup() -> Void {
        
        self.groupPrefix = nil;
        self.group = NSNumber(value: false as Bool)
    
    }
    
    //MARK: static implementation section
    
    static let _REGEXP = "(\\w+[-@/])(\\w+)"
    
    static func isSectionAware( _ key:KeyEntity ) -> Bool {
        
        let predicate = NSPredicate(format: "SELF.mnemonic MATCHES %@", _REGEXP)
        
        return predicate.evaluate(with: key)
    }
    
    static func getSectionPrefix( _ key:KeyEntity, checkIfIsSectionAware:Bool ) -> NSRange {
    
        var result = NSMakeRange(NSNotFound, 0)
        
        if( !checkIfIsSectionAware || KeyEntity.isSectionAware(key) ) {
        
            do {
                let pattern = try NSRegularExpression(pattern: _REGEXP, options:.caseInsensitive)

                let length = key.mnemonic.characters.count
                
                if let match:NSTextCheckingResult =
                            pattern.firstMatch( in: key.mnemonic,
                                                        options:.reportCompletion,
                                                        range:NSMakeRange(0, length))
                {

                    result = match.rangeAt(1)
                            
                }

                
            }
            catch let error as NSError {
                
                print ("Error: \(error.domain)")
            }
        
        }
    
        return result
    
    }
    
    static func sectionNameFromPrefix( _ prefix:String?, trim:Bool ) -> String?  {
        
        guard let p = prefix else {
            return nil;
        }
        
        var pp:String = p
        
        if( trim ) {
            
            pp = p.trimmingCharacters(in: CharacterSet.whitespacesAndNewlines)
        }
        
        if( pp.isEmpty ) {
            return nil;
        }
        
        let index = pp.characters.index(before: pp.endIndex)
        
        let result = pp.substring(to: index)
        
        return result;
    
    }

    static func createSection(  _ groupKey:String,
                                groupPrefix:String,
                                inContext context:NSManagedObjectContext) -> KeyEntity?
    {
    
        guard let entity = NSEntityDescription.entity(forEntityName: "KeyInfo", in:context) else {
            return nil
        }

        guard let entityName = entity.name else {
            return nil
        }
            
        guard let cloned = NSEntityDescription.insertNewObject(forEntityName: entityName, into:context) as? KeyEntity else {
            return nil
        }
        
        cloned.setValue("nil", forKey: "password")
        cloned.setValue("nil", forKey: "username")
    
        cloned.group = NSNumber(value: false as Bool)
        cloned.mnemonic = groupKey;
        cloned.groupPrefix = groupPrefix;

        return cloned
    }
    
    // Copy all passwords to KeyChain and invalidate related password fields
    static func copyPasswordsToKeychain( _ context:NSManagedObjectContext ) {
        
        let fetchRequest = NSFetchRequest<NSFetchRequestResult>(entityName: "KeyInfo")
                
        do {

            if let result = try context.fetch(fetchRequest) as? [KeyEntity] {
             
                result.forEach({ (kk:KeyEntity) -> () in

                    copyPasswordToKeychain( kk )
                    
                })
            }
        }
        catch let error as NSError {
            print( "error perform fetch \(error.userInfo)")
        }
        
    }
    
    // Copy single password to KeyChain and invalidate related password field
    static func copyPasswordToKeychain( _ kk:KeyEntity ) {
        
        
        guard   let data = kk.value(forKey: "password") as? Data,
            let key = kk.value(forKey: "mnemonic") as? String else {
                return
        }

        do {
            

            if let password = NSString(data:data, encoding:String.Encoding.utf8.rawValue) as? String {
                
                print("key:\(key) password:\(password)")
                
                try Keychain(service: AccountCredential.sharedCredential.bundleId)
                    .accessibility(.whenUnlocked)
                    .set(data, key: key)
                
            }
        
        }
        catch let error as NSError {
            print( "error putting in keychain key \(key) - \(error.userInfo)")
        }
        
        
        
    }
    
    // MARK: implementation
    
    func isSection() -> Bool {
        return self.groupPrefix != nil && (self.group == nil || self.group?.boolValue == false)
    }
    
    func isGrouped() -> Bool {
        return (self.groupPrefix != nil && self.group != nil && self.group!.boolValue)
    }
    
    func isEqualForImport( _ object:AnyObject? ) -> Bool {
    
        guard let o = object  else {
            return false
        }
        
        
        var k1:String?
    
        if ( o is KeyEntity ) {
    
            k1 = o.mnemonic;
    
        }
        else if( o is NSDictionary ) {
    
            let d = o as! NSDictionary
            
            k1 = d.value(forKey: KeyEntity._MNEMONIC) as? String
    
        }
        else if( o is String ) {
    
            k1 = o as? String
    
        }
    
        
        return self.mnemonic == k1
        
        
    
    }
    
    
    override func awakeFromFetch() {
    
        self.setPrimitiveValue(NSNumber(value: false as Bool), forKey:KeyEntity._IS_NEW)
    
    }
    
    
    override func didSave() {
        
        self.setPrimitiveValue(NSNumber(value: false as Bool), forKey:KeyEntity._IS_NEW)
    }
    
    //MARK: Serialization
    
    let sortPredicate = { (A:(key:String, _:NSAttributeDescription), B:(key:String, _:NSAttributeDescription)) -> Bool in
        return A.key < B.key
    }
    
    let filterPredicate = { (E:(key:String, _:NSAttributeDescription)) -> Bool in
        return E.key != KeyEntity._IS_NEW
    }
    
    func toDictionary( _ target:NSMutableDictionary? ) -> NSDictionary? {
    
        guard let t = target else {
            return target;
        }
    
        self.entity.attributesByName
            .filter( filterPredicate )
            .sorted( by: sortPredicate )
            .forEach { (key:String, _:NSAttributeDescription) -> () in
            
                    if let v = self.value(forKey: key) {
                        t.setObject(v, forKey: key as NSCopying)
                    }
                }
            
                
        return t;
    }
    
    func fromDictionary(_ source:NSDictionary?) {
    
        guard let s = source else {
            return
        }
    
        self.entity.attributesByName
            .filter( filterPredicate )
            .sorted( by: sortPredicate )
            .forEach { (key:String, _:NSAttributeDescription) -> () in
        
                if let value = s.value(forKey: key) {
                    self.setValue(value, forKey: key)
                }
            
        }
    }

}
